### PR TITLE
Modify data loading to handle non-existing masks

### PR DESCRIPTION
**Pull request by PhilipMathieu**
This modification generates an empty (black) mask if the mask image is not present. I also included a new notebook that loads an image and verifies bit depths (which could be convenient for debugging)

### CODE REVIEW

**Comment by PhilipMathieu** [comment]
This is not ready to merge yet. Specifically:
- [x] Preprocessing script needs to be modified to account for all-0 images (rather than allowing NaNs to appear)
- [x] BCELoss function needs to have a small error term added to eliminate log(0) returning NaNs
- [x] Need to add a try except in the dataloader to catch indexing errors

**Comment by PhilipMathieu** [comment]
This has been updated. It fixes #6 but reveals a new issue (the model does not train with imbalanced classes - it just learns to guess 0 for everything)

File: src/train.py [hunk 1]
```diff
@@ -176,16 +180,17 @@ def train_model(
 def get_args():
     parser = argparse.ArgumentParser(description='Train the UNet on images and target masks')
     parser.add_argument('--epochs', '-e', metavar='E', type=int, default=20, help='Number of epochs')
-    parser.add_argument('--batch-size', '-b', dest='batch_size', metavar='B', type=int, default=16, help='Batch size')
+    parser.add_argument('--batch-size', '-b', dest='batch_size', metavar='B', type=int, default=32, help='Batch size')
     parser.add_argument('--learning-rate', '-l', dest='lr', metavar='LR', type=float, default=1e-4, help='Learning rate')
     parser.add_argument('--load', '-f', type=str, default=False, help='Load model from a .pth file')
-    parser.add_argument('--scale', '-s', type=float, default=0.5, help='Downscaling factor of the images')
+    parser.add_argument('--scale', '-s', type=float, default=1.0, help='Downscaling factor of the images')
```
Comment (rodroadl) [inline src/train.py]: Why are we setting this to 1.0?

File: src/utils/data_loading.py
Comment (rodroadl) [inline src/utils/data_loading.py:1]: Whenver you get a chance, could you explain why we need pos_weight and why we are using self.stats as pos_weight?
Not a high priority.

**Review by rodroadl — APPROVED** [review]
Looks good. Left some comments.

File: src/utils/data_loading.py
Comment (rodroadl) [reply src/utils/data_loading.py:1]: > This has been updated. It fixes #6 but reveals a new issue (the model does not train with imbalanced classes - it just learns to guess 0 for everything)

I assume pos_weight is to sovle this by trading off recall and precision, but still not working. Am I correct?

Comment (PhilipMathieu) [reply src/utils/data_loading.py:1]: pos_weight adjusts the value of BCELoss to account for the difference in # of positive vs non-positive examples. From the BCEWithLogitsLoss docs:
"For example, if a dataset contains 100 positive and 300 negative examples of a single class, then pos_weight for the class should be equal to 300/100 = 3. The loss would act as if the dataset contains 3×100=300 positive examples."

File: src/train.py (hunk 1 above)
Comment (PhilipMathieu) [reply src/train.py]: Because the training images are 64x64 now, so we do not need to scale them

File: src/utils/data_loading.py [hunk 2]
```diff
@@ -67,6 +88,11 @@ def __init__(self, images_dir: str, dem_dir: str, mask_dir: str, scale: float =
     def __len__(self):
         return len(self.ids)
     
+    def _process_stats(self):
```
Comment (PhilipMathieu) [inline src/utils/data_loading.py:91]: This function uses the information from the ESRI json file to calculate the correct pos_weight for BCEWithLogitsLoss

File: src/train.py (hunk 1 above)
Comment (rodroadl) [reply src/train.py]: Got it. I thoguth it was rate for down convolution layer. Good catch.

File: src/utils/data_loading.py
Comment (rodroadl) [reply src/utils/data_loading.py:1]: Alright, but it is not resloving the model does not train with imbalanced classes - it just learns to guess 0 for everything.

Comment (PhilipMathieu) [reply src/utils/data_loading.py:1]: Correct, unfortunately
